### PR TITLE
Refactor type and range validation in configuration update process

### PR DESCRIPTION
**Pull request by winterhazel**
### Description

The configuration update process mixes type validation with range validation in `com.cloud.configuration.ConfigurationManagerImpl#validateConfigurationValue`. To improve this process, this PR splits these validations into two methods (`validateValueType` and `validateValueRange`) and cleans up the code.

Furthermore, this PR also fixes an issue that allows non-numeric values to be inserted into `Double` configurations, such as `zone.virtualnetwork.ipv6subnet.capacity.notificationthreshold`.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [X] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### How Has This Been Tested?

I manually verified the behavior of trying to update configurations of the following types, and compared them with the previous version to ensure the result was expected:

- `String` - `quota.currency.symbol` and `alert.email.addresses` (that is, one that does not have `null` as a valid value and another that does);
- `Boolean` - `admin.is.allowed.to.deploy.anywhere`;
- `Integer` - `storage.cleanup.interval`. (the validations for this type also apply to `Short` and `Long`);
- `Float` - `cpu.overprovisioning.factor` (also applies to `Double`).

| Number | Test | Result | Expected result (Y/N)? |
| - | - | - | - |
| 1 | Value `"null"` in a `String` that does not have `null` as a valid value | The value is updated in the database, but the default value is used | Y |
| 2 | Value `"1"` in a `String` that does not have `null` as a valid value | Success | Y |
| 3 | Value `"1.1"` in a `String` that does not have `null` as a valid value | Success | Y |
| 4 | Value `"test"` in a `String` that does not have `null` as a valid value | Success | Y |
| 5 | Value `"null"` in a `String` that has `null` as a valid value | Success | Y |
| 6 | Value `"1"` in a `String` that has `null` as a valid value | Success | Y |
| 7 | Value `"1.1"` in a `String` that has `null` as a valid value | Success | Y |
| 8 | Value `"test"` in a `String` that has `null` as a valid value | Success | Y |
| 9 | Value `"null"` in a `Boolean` | Exception | Y |
| 10 | Value `"1"` in a `Boolean` | Exception | Y |
| 11 | Value `"1.1"` in a `Boolean` | Exception | Y |
| 12 | Value `"test"` in a `Boolean` | Exception | Y |
| 13 | Value `"true"` in a `Boolean` | Success | Y |
| 14 | Value `"false"` in a `Boolean` | Success | Y |
| 15 | Value `"fALse"` in a `Boolean` | Exception | Y |
| 16 | Value `"null"` in an `Integer` | Exception | Y |
| 17 | Value `"1"` in an `Integer` | Success | Y |
| 18 | Value `"1.1"` in an `Integer` | Exception | Y |
| 19 | Value `"test"` in an `Integer` | Exception | Y |
| 20 | Value `"null"` in a `Float` | Exception | Y |
| 21 | Value `"1"` in a `Float` | Success | Y |
| 22 | Value `"1.1"` in a `Float` | Success | Y |
| 23 | Value `"test"` in a `Float` | Exception | Y |

### CODE REVIEW

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 1]
```diff
@@ -682,9 +682,8 @@ public boolean stop() {
     @DB
     public String updateConfiguration(final long userId, final String name, final String category, String value, final String scope, final Long resourceId) {
         final String validationMsg = validateConfigurationValue(name, value, scope);
-
         if (validationMsg != null) {
-            logger.error("Invalid configuration option, name: " + name + ", value:" + value);
+            logger.error("Invalid value [%s] for configuration [%s] due to [%s].", value, name, validationMsg);
```
Comment (hsato03) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:686]:
```suggestion
            logger.error("Invalid value [{}] for configuration [{}] due to [{}].", value, name, validationMsg);
```

**Comment by winterhazel** [comment]
@blueorangutan package

**Comment by winterhazel** [comment]
@blueorangutan package

**Comment by winterhazel** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -1211,132 +1215,142 @@ protected String validateConfigurationValue(final String name, String value, fin
...
-        value = value.trim();
         try {
-            if (overprovisioningFactorsForValidation.contains(name) && Float.parseFloat(value) <= 0f) {
-                final String msg = name + " should be greater than 0";
-                logger.error(msg);
-                throw new InvalidParameterValueException(msg);
+            if (type == Boolean.class) {
+                return value.equals("true") || value.equals("false");
+            } else if (type == Integer.class || type == Long.class || type == Short.class) {
+                Integer.parseInt(value);
```
Comment (JoaoJandre) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1253]: If the value is a long, will Integer.parseInt() be able to parse it? Shouldn't we use Long.parseLong()?

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 3]
```diff
@@ -1211,132 +1215,142 @@ protected String validateConfigurationValue(final String name, String value, fin
...
-                if (UnmanagedVMsManager.ThreadsOnMSToImportVMwareVMFiles.key().equalsIgnoreCase(name) || UnmanagedVMsManager.ThreadsOnKVMHostToImportVMwareVMFiles.key().equalsIgnoreCase(name)) {
-                    if (val > 10) {
-                        throw new InvalidParameterValueException("Please enter a value between 0 and 10 for the configuration parameter: " + name + ", -1 will disable it");
-                    }
+            }
+            if (UnmanagedVMsManager.ThreadsOnMSToImportVMwareVMFiles.key().equalsIgnoreCase(name) ||
+                    UnmanagedVMsManager.ThreadsOnKVMHostToImportVMwareVMFiles.key().equalsIgnoreCase(name)) {
+                if (val > 10) {
+                    return String.format("Please enter a value between 0 and 10 for the configuration parameter: [%s]. -1 will disable it.", name);
                 }
```
Comment (JoaoJandre) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1298]:
```suggestion
                    UnmanagedVMsManager.ThreadsOnKVMHostToImportVMwareVMFiles.key().equalsIgnoreCase(name)) {
                if (val < -2 || val > 10) {
                    return String.format("Please enter a value between 0 and 10 for the configuration parameter: [%s]. -1 will disable it.", name);
                }
```

Comment (hsato03) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1298]: Shouldn't the returned message be updated as well?

**Review by hsato03 — APPROVED** [review]
LGTM. didn't test

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 3 above)
Comment (hsato03) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1298]: Shouldn't the returned message be updated as well?

**Comment by JoaoJandre** [comment]
Hey @winterhazel

I did some testing, overall it looks good, but I've found one case that I think we should address in this PR.

| Number | Config | Value | Result |
| --- | --- | --- | --- |
| 1 | storage.cleanup.interval | oasij | Error |
| 2 | storage.cleanup.interval | . | Error |
| 3 | storage.cleanup.interval | ( | Error |
| 4 | storage.cleanup.interval | 102938120381209381203981239128 | Error |
| 5 | storage.cleanup.interval |  | Error |
| 6 | storage.cleanup.interval | "" | Error |
| 7 | storage.cleanup.interval | 86400 | Ok |
| 8 | cpu.overprovisioning.factor | a | Error |
| 9 | cpu.overprovisioning.factor | ^ | Error |
| 10 | cpu.overprovisioning.factor | 1.1.1 | Error |
| 11 | cpu.overprovisioning.factor | 10928091283012830123810293 | OK |
| 12 | cpu.overprovisioning.factor | 9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999 | OK, but this should not be |
| 13 | admin.is.allowed.to.deploy.anywhere | 1 | Error |
| 14 | admin.is.allowed.to.deploy.anywhere | 0 | Error |
| 15 | admin.is.allowed.to.deploy.anywhere | a | Error |
| 16 | admin.is.allowed.to.deploy.anywhere | FALSE | Error |
| 17 | admin.is.allowed.to.deploy.anywhere | TRUE | Error |
| 18 | admin.is.allowed.to.deploy.anywhere | true | OK |
| 19 | admin.is.allowed.to.deploy.anywhere | false | OK |

Test 12 was sucessful, but the number informed is bigger than the maximum value of Floats and Doubles. Double.parseDouble() returns Infinity if the number being parsed would overflow. We should not allow this config value, we should return an error, like what is done in test 4.

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 2 above)
Comment (winterhazel) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1253]: You're right, it won't if the value exceeds +/- 2147483647. I'll adjust it.

**Comment by winterhazel** [comment]
Thanks for testing @JoaoJandre. I addressed the issue you pointed out.

**Comment by winterhazel** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan LLtest

**Comment by winterhazel** [comment]
@blueorangutan package

**Comment by JoaoJandre** [comment]
Merging based on approvals and manual testing https://github.com/apache/cloudstack/pull/9107#issuecomment-2305311036

**Pull request merged**